doubleml/double_ml_selection.py

```diff
@@ -0,0 +1,607 @@
+            mu_hat_d0['targets'] = np.hstack(mu_d0['targets'])
+            pi_hat_d0['targets'] = np.hstack(pi_d0['targets'])
+            p_hat_d0['targets'] = np.hstack(p_d0['targets'])
+            s_0 = np.hstack(s_0)
+            y_0 = np.hstack(y_0)
+            dcontrol = np.hstack(dcontrol)
+
+        p_hat_d0['preds'] = _trimm(p_hat_d0['preds'], self._trimming_rule, self._trimming_threshold)
+        p_hat_d1['preds'] = _trimm(p_hat_d1['preds'], self._trimming_rule, self._trimming_threshold)
+
+        psi_a, psi_b = self._score_elements(dtreat, dcontrol, mu_hat_d1['preds'],
+                                            mu_hat_d0['preds'], pi_hat_d1['preds'],
+                                            pi_hat_d0['preds'],
+                                            p_hat_d1['preds'], p_hat_d0['preds'],
+                                            s_0, s_1, y_0, y_1)
+
+        psi_elements = {'psi_a': psi_a,
+                        'psi_b': psi_b}
+
+        preds = {'predictions': {'ml_mu_d0': mu_hat_d0['preds'],
+                                 'ml_mu_d1': mu_hat_d1['preds'],
+                                 'ml_pi_d0': pi_hat_d0['preds'],
+                                 'ml_pi_d1': pi_hat_d1['preds'],
+                                 'ml_p_d0': p_hat_d0['preds'],
+                                 'ml_p_d1': p_hat_d1['preds']},
+                 'targets': {'ml_mu_d0': mu_hat_d0['targets'],
+                             'ml_mu_d1': mu_hat_d1['targets'],
+                             'ml_pi_d0': pi_hat_d0['targets'],
+                             'ml_pi_d1': pi_hat_d1['targets'],
+                             'ml_p_d0': p_hat_d0['targets'],
+                             'ml_p_d1': p_hat_d1['targets']},
+                 'models': {'ml_mu_d0': mu_hat_d0['models'],
+                            'ml_mu_d1': mu_hat_d1['models'],
+                            'ml_pi_d0': pi_hat_d0['models'],
+                            'ml_pi_d1': pi_hat_d1['models'],
+                            'ml_p_d0': p_hat_d0['models'],
+                            'ml_p_d1': p_hat_d1['models']}
+                 }
+
+        return psi_elements, preds
+
+    def _score_elements(self, dtreat, dcontrol, mu_d1, mu_d0,
+                        pi_d1, pi_d0, p_d1, p_d0, s_0, s_1, y_0, y_1):
+        # psi_a
+        psi_a = -1
+
+        # psi_b
+        if self._normalize_ipw:
+            weight_treat = sum(dtreat) / sum((dtreat * s_1) / (pi_d1 * p_d1))
+            weight_control = sum(dcontrol) / sum((dcontrol * s_0) / (pi_d0 * p_d0))
+
+            psi_b1 = weight_treat * ((dtreat * s_1 * (y_1 - mu_d1)) / (p_d1 * pi_d1)) + mu_d1
+            psi_b0 = weight_control * ((dcontrol * s_0 * (y_0 - mu_d0)) / (p_d0 * pi_d0)) + mu_d0
+
+        else:
+            psi_b1 = (dtreat * s_1 * (y_1 - mu_d1)) / (p_d1 * pi_d1) + mu_d1
+            psi_b0 = (dcontrol * s_0 * (y_0 - mu_d0)) / (p_d0 * pi_d0) + mu_d0
+
+        psi_b = psi_b1 - psi_b0
```

Comment:
You are absolutely right, I don't know why I didn't notice this. I asked Lukas Laffers about using just one sample splitting and he confirmed that it should be OK, so I will rewrite the code to use just one procedure and that should also help me to eliminate this issue. I will also ask about the variance estimation.

Reply:
Yes, I already started rewriting the code so that it would all match the order of the input data. I think I should be able to get back to it this week.